### DISCUSSION TITLE
Increase max session time

### DISCUSSION THREAD

**Discussion by azhinu** [gravitational/teleport#30213] [Q&A] [answered]
Hello everybody. I'm trying to increase session time, but with no success. 
Firstly, I tried to increase `max_session_ttl` in role options, but nothing changed. Then I changed `auth_service.authentication.default_session_ttl`, but still no success. Teleport still require login every 12 hours.
How can I increase time that i haven't to relogin teleport?

**Comment by webvictim** [comment]
You will get the lowest value for `max_session_ttl` which is set across all your roles, so make sure that `max_session_ttl` is set to > 12 hours in _EVERY_ role that you have.

**Reply by azhinu** [reply]
All roles have `max_session_ttl` set to 230h

**Reply by webvictim** [reply, marked as answer]
Try setting everything to `30h` and see if that works. Teleport has an upper limit of 30 hours on session length.

**Reply by azhinu** [reply]
Is there a way to bypass this restriction?

**Reply by webvictim** [reply]
You could fork the code, edit it and recompile Teleport. The limit is there for security, as issuing a certificate allowing access for over a week based on one login is undesirable in most cases.

**Reply by kelvich** [reply]
Can you let your users decide on the desired tradeoff between risk and amount of inconvenience? I get it when products forbid some outright bad configurations, but requiring login every 30 hours is not an industry standard by any means, e.g., none of the major clouds has such a limit. In my opinion, this is a sign of health issues with your product management—somebody on the security side has overreaching influence and imposes decisions that result in infinitesimal security gains at the cost of major user discomfort.

**Reply by dmelo** [reply]
I have set all my roles to have `max_session_ttl` set to 30h, but my certificate is valid only for 12h. Is there anywhere else that I have to set TTL? 

In the end, I want users who login using Web UI to have more time before Teleport asks to login again. Or perhaps a way to trust the device (the browser the user is using). Is there any way to do this?

I'm asking about this questions again, because it has been a while since @azhinu posted the original question and I see that Teleport evolves quickly.

**Reply by webvictim** [reply]
Teleport defaults to issuing certs with `tsh login` which are valid for 12 hours. You can request a longer-lived cert (up to 30 hours) using `tsh login --ttl 1800` - 1800 minutes is 30 hours.

To make this the default, edit `/etc/teleport.yaml` on machines running the Teleport `auth_service`, make this change and restart:
```
auth_service:
  authentication:
    default_session_ttl: 30h
```

This should give you 30 hour certs with `tsh login` as long as `max_session_ttl` is set high enough in your role.

(for Teleport Enterprise Cloud customers, you can do this using `tctl edit cluster_auth_preference`):
```
spec:
  default_session_ttl: 30h
```

**Reply by webvictim** [reply]
I'm going to post a second reply to talk about the web UI specifically, as it doesn't work the same way as `tsh login` does.

Internally, the web UI uses a user certificate which is referenced in the browser with a bearer token which is valid for around ~10 minutes. This bearer token is refreshed every ~8 minutes using a background loop, but ONLY as long as the user's session in the web UI remains active and the browser is able to make requests to the cluster.

If users are being logged out before the default 12 hour expiry hits, this is most likely because the bearer token isn't being refreshed. When this happens and the bearer token expires without being refreshed, the user is logged out.

This can happen for a number of reasons:
- the user closes their laptop and the browser can't refresh the token
- the user disconnects from wifi or changes connectivity to use a VPN/not use a VPN
- the user changes tabs away from Teleport and the browser puts the tab to sleep to save memory
- an addon puts the tab to sleep
- various other reasons

Essentially as long as the browser is open and the bearer token is being refreshed, the user's session will stay active. However, if you are expecting a user to be able to log in, use the Teleport web UI for an hour, close their laptop for a few hours, come back and not have to log into the web UI again, that will never be doable.

For longer-lived sessions and use cases like the one I described above, I highly recommend that users use [Teleport Connect](https://goteleport.com/docs/connect-your-client/teleport-connect/) or `tsh login` in that scenario, where you can get longer-lived certs which are stored locally.

**Reply by dmelo** [reply]
@webvictim , thanks for the tip about Teleport Connect. I'm trying it here, but when I try to launch a Web App, it redirects to the browser in a position where I have to login again on the browser. So it doesn't help if the goal is to reduce logins, for Web Apps. Am I missing something?

**Reply by webvictim** [reply]
No, you're right - Teleport Connect just hands off to the browser for logging into applications. It will work to make SSH/Kubernetes/Databases easier, though.

If the goal is to reduce friction from repeated logins, have you tried adding a passwordless login to Teleport? That way you don't need a username, password, and second factor - you just need to use something which supports passkeys like most modern browsers, password managers, Touch ID (Mac), Windows Hello, or a physical device like a Yubikey, etc.

**Reply by OneideLuizSchneider** [reply]
@webvictim is this still valid for the latest versions (16.x)? (TTL up to 30h)

**Reply by webvictim** [reply]
Yes.

**Reply by beleyj-ss** [reply]
@webvictim For the Web GUI, is there a way to force the user to log back in after so many hours? Based on your description above, it sounds like if the browser session is active (aka the tab is open), it will continue to stay logged in for a infinite amount of time.

**Reply by webvictim** [reply]
@beleyj-ss The upper limit on time that a user can be logged into the web UI is still dictated by the maximum session TTL set for each role they have assigned:

```yaml
spec:
  options:
    max_session_ttl: 8h
```

If the user has multiple roles assigned, their maximum session length is capped at the lowest value of `max_session_ttl` across all their roles.

Once they hit the maximum session length, they'll be logged out and required to re-authenticate.

**Reply by KamiSamaD** [reply]
@webvictim  Hi, when I set ` auth_service.authentication.default_session_ttl: 15h` and all roles `max_session_ttl `is 30h ,  but it can not work, still 12h ttl

**Reply by KamiSamaD** [reply]
`tsh login --ttl 1800`  is worked, but default set not work. teleport version is v1.17.3

**Reply by kriegster108** [reply]
I dont get why we cant have longer lived credentials in the browser. This is a normal feature for 99% of modern web applications, you shouldnt need to have to keep your laptop online or browser tab open to stay logged into a web application.

It's pretty archaic to say that its impossible for not allowing your customers to configure TTL for tokens being issued by teleport so we can provide our internal users a better experience.

My team having to tell everyone in our org they have to install a desktop app to get a better experience is overkill when the webui that exists already is sufficient to do the job. 

So if you use Connect, CLI, SSH to have a longer lived credential, but when it comes to the webui its a hard "impossible"? Makes no sense to me.

**Reply by zmb3** [reply]
You might be interested in https://github.com/gravitational/teleport/pull/61103
